src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs

```diff
@@ -0,0 +1,512 @@
+                    {
+                        case (test: BoundDagNonNullTest _, sense: true):
+                        case (test: BoundDagExplicitNullTest _, sense: false):
+                            // not-null tests are implicitly incorporated into a recursive pattern
+                            break;
+                        default:
+                            return null;
+                    }
+                }
+
+                string deconstruction = null;
+                var properties = new Dictionary<Symbol, string>();
+                bool needsPropertyString = false;
+
+                foreach (var eval in evaluations)
+                {
+                    switch (eval)
+                    {
+                        case BoundDagDeconstructEvaluation e:
+                            var method = e.DeconstructMethod;
+                            int extensionExtra = method.RequiresInstanceReceiver ? 0 : 1;
+                            int count = method.Parameters.Length - extensionExtra;
+                            var subpatterns = new ArrayBuilder<string>(count);
+                            for (int j = 0; j < count; j++)
+                            {
+                                var elementTemp = new BoundDagTemp(e.Syntax, method.Parameters[j + extensionExtra].Type, e, j);
+                                var newPattern = SamplePatternForTemp(elementTemp, constraintMap, evaluationMap, requireExactType: false);
+                                subpatterns.Add(newPattern);
+                            }
+                            var result = "(" + string.Join(", ", subpatterns) + ")";
+                            if (deconstruction != null && needsPropertyString)
+                            {
+                                deconstruction = deconstruction + " { }";
+                                needsPropertyString = properties.Count != 0;
+                            }
+
+                            deconstruction = (deconstruction is null) ? result : deconstruction + " and " + result;
+                            needsPropertyString |= count == 1;
+                            break;
+                        case BoundDagFieldEvaluation e:
+                            {
+                                var subInput = new BoundDagTemp(e.Syntax, e.Field.Type, e);
+                                var subPattern = SamplePatternForTemp(subInput, constraintMap, evaluationMap, false);
+                                properties.Add(e.Field, subPattern);
+                            }
+                            break;
+                        case BoundDagPropertyEvaluation e:
+                            {
+                                var subInput = new BoundDagTemp(e.Syntax, e.Property.Type, e);
+                                var subPattern = SamplePatternForTemp(subInput, constraintMap, evaluationMap, false);
+                                properties.Add(e.Property, subPattern);
```

Comment:
>properties.Add(e.Property, subPattern) [](start = 32, length = 38)

Does this handle duplicated properties?
```C#
_ = o switch { C { P: 1, P: 1 } => 1 };
```

Reply:
Yes, but since the duplicate would be removed during dag construction, it won't be duplicated in a resulting explained pattern. E.g `switch { not { P: 1, P: 1 } }` would say that `{P: 1}` is not covered.

---
In reply to: [443811618](https://github.com/dotnet/roslyn/pull/44702#discussion_r443811618) [](ancestors = 443811618)

Reply:
Added a test.

---
In reply to: [443825104](https://github.com/dotnet/roslyn/pull/44702#discussion_r443825104) [](ancestors = 443825104,443811618)